polars/polars-time/src/windows/groupby.rs

```diff
@@ -233,16 +233,14 @@ pub(crate) fn groupby_values_iter_full_lookbehind<'a>(
     tu: TimeUnit,
     tz: Option<impl PolarsTimeZone + 'a>,
     start_offset: usize,
-) -> impl Iterator<Item = (IdxSize, IdxSize)> + TrustedLen + 'a {
+) -> impl Iterator<Item = PolarsResult<(IdxSize, IdxSize)>> + TrustedLen + 'a {
     debug_assert!(offset.duration_ns() >= period.duration_ns());
     debug_assert!(offset.negative);
-    fn add<T: PolarsTimeZone>(tu: TimeUnit) -> fn(&Duration, i64, Option<&T>) -> PolarsResult<i64> {
```

Comment:
@MarcoGorelli this was a double function call and therefore more expensive than it needs to be. 

Given that we already have a dynamic function call here (which is expensive, but worth it because of the many options in time related logic), we should probably factor out the time zone logic as well. 

We could  make an `add_ns` and an `add_ns_tz`. As long as we ensure the function signatures remain the same, we can use dynamic dispatch and we can get some performance boosts.